### PR TITLE
feat: Add collab liquidation

### PR DESCRIPTION
**Pull request by holzeis**
Analog to the expired positions this is adding an optimistic approach to close liquidated positions collaboratively.

If the ask or the bid price gets within a range of 10% to the liquidation price, an async order with order reason `Liquidated` is automatically created and the user notified about the liquidated positions. The user will get daily reminders for a week to collab close the liquidated position.

This means with an average entry price of $30,000 the user gets liquidated on a short position at $37,500 - $3,750 = $33,750.

https://github.com/get10101/10101/assets/382048/0ab2da62-ffdf-415c-8382-1fe1ddba4171

fixes https://github.com/get10101/meta/issues/407

### CODE REVIEW

**Comment by holzeis** [comment]
Should we show the margin call in the app?

**Comment by holzeis** [comment]
The order reason is shown in the order history if the order reason is different to "manual"

image

**Review by bonomat — CHANGES_REQUESTED** [review]
I'm afraid I put you onto the wrong track here. The way how this is used and called at the moment is not a `margin call` and I also put you onto the wrong track on how it is calculated.
I'm sorry but I would rather not merge this as is then having a wrong terminology introduced which might confuse us and traders down the road.

Here is what we should do: 

Implement a maintenance margin and derive the liquidation price from this maintenance margin. 
I would then simply replace the current `Liquidation Price` with the calculation from below and optionally add a `Bankruptcy Price`: this is where the user loses everything. 

Here a full example: 

### Assumptions:
- Initial price of Bitcoin (BTC): $10,000
- Contract size: $10,000
- Initial margin requirement: 50% (leverage of 2x)
- Maintenance margin requirement: 10% (this is quite high but maybe OK for us)

### Scenario 1: Opening a Long Position
1. **Opening a Long Position**:
    - Contract size: $10,000 
    - Initial Leverage: 2x (you need to deposit 50% of the total contract size as the initial margin).
    - Initial Margin = 50% * $10,000 = $5,000
    - The initial margin in BTC using the current price of Bitcoin:
      BTC Initial Margin = $5,000 / $10,000 per BTC = 0.5 BTC

2. **Liquidation Price Calculation**:
    - This is the formula to calculate the liquidation price using maintenance margin: `(price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))`
    - If maintenance_margin = 0%: 
      - Liquidation Price = (price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))
        Liquidation Price = ($10,000 * 2) / (2 + 1 - (0.1 * 0))
        Liquidation Price = $20,000 / (3 - 0)
        Liquidation Price = $20,000 / 3
        Liquidation Price ≈ $6,666.67 
    - If maintenance_margin = 10%:
      - Liquidation Price = (price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))
        Liquidation Price = ($10,000 * 2) / (2 + 1 - (0.1 * 2))
        Liquidation Price = $20,000 / (3 - 0.2)
        Liquidation Price = $20,000 / 2.8
        Liquidation Price ≈ $7,142.86
      - --> meaning, the user gets liquidated earlier

### Scenario 2: Opening a Short Position
1. **Opening a Short Position**:
    - Contract size: $10,000
    - Initial Leverage: 2x (you need to deposit 50% of the total contract size as the initial margin).
    - Initial Margin = 50% * $10,000 = $5,000
    - The initial margin in BTC using the current price of Bitcoin:
      BTC Initial Margin = $5,000 / $10,000 per BTC = 0.5 BTC

2. **Liquidation Price Calculation**:
    - If maintenance_margin = 0%:
      - Liquidation Price = (price_opening * leverage) / (leverage - 1 + (maintenance_margin * leverage))
        Liquidation Price = ($10,000 * 2) / (2 - 1 + (0.1 * 0))
        Liquidation Price = $20,000 / (2 - 1)
        Liquidation Price = $20,000 / 1
        Liquidation Price = $20,000
        Liquidation Price = $20,000 per BTC
   - If `maintenance_margin = 10%`:
       - Liquidation Price = (price_opening * leverage) / (leverage - 1 + (maintenance_margin * leverage))
         Liquidation Price = ($10,000 * 2) / (2 - 1 + (0.1 * 2))
         Liquidation Price = $20,000 / (2 - 1 + 0.2)
         Liquidation Price = $20,000 / (1 + 0.2)
         Liquidation Price = $20,000 / 1.2
         Liquidation Price ≈ $16,666.67
       - --> meaning, the user gets liquidated earlier

**Comment by holzeis** [comment]
> I'm afraid I put you onto the wrong track here. The way how this is used and called at the moment is not a `margin call` and I also put you onto the wrong track on how it is calculated. I'm sorry but I would rather not merge this as is then having a wrong terminology introduced which might confuse us and traders down the road.
> 
> Here is what we should do:
> 
> Implement a maintenance margin and derive the liquidation price from this maintenance margin. I would then simply replace the current `Liquidation Price` with the calculation from below and optionally add a `Bankruptcy Price`: this is where the user loses everything.
> 
> Here a full example:
> ### Assumptions:
> 
>     * Initial price of Bitcoin (BTC): $10,000
> 
>     * Contract size: $10,000
> 
>     * Initial margin requirement: 50% (leverage of 2x)
> 
>     * Maintenance margin requirement: 10% (this is quite high but maybe OK for us)
> 
> 
> ### Scenario 1: Opening a Long Position
> 
>     1. **Opening a Long Position**:
>        
>        * Contract size: $10,000
>        * Initial Leverage: 2x (you need to deposit 50% of the total contract size as the initial margin).
>        * Initial Margin = 50% * $10,000 = $5,000
>        * The initial margin in BTC using the current price of Bitcoin:
>          BTC Initial Margin = $5,000 / $10,000 per BTC = 0.5 BTC
> 
>     2. **Liquidation Price Calculation**:
>        
>        * This is the formula to calculate the liquidation price using maintenance margin: `(price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))`
>        * If maintenance_margin = 0%:
>          
>          * Liquidation Price = (price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))
>            Liquidation Price = ($10,000 * 2) / (2 + 1 - (0.1 * 0))
>            Liquidation Price = $20,000 / (3 - 0)
>            Liquidation Price = $20,000 / 3
>            Liquidation Price ≈ $6,666.67
>        * If maintenance_margin = 10%:
>          
>          * Liquidation Price = (price_opening * leverage) / (leverage + 1 - (maintenance_margin * leverage))
>            Liquidation Price = ($10,000 * 2) / (2 + 1 - (0.1 * 2))
>            Liquidation Price = $20,000 / (3 - 0.2)
>            Liquidation Price = $20,000 / 2.8
>            Liquidation Price ≈ $7,142.86
>          * --> meaning, the user gets liquidated earlier
> 
> 
> ### Scenario 2: Opening a Short Position
> 
>     1. **Opening a Short Position**:
>        
>        * Contract size: $10,000
>        * Initial Leverage: 2x (you need to deposit 50% of the total contract size as the initial margin).
>        * Initial Margin = 50% * $10,000 = $5,000
>        * The initial margin in BTC using the current price of Bitcoin:
>          BTC Initial Margin = $5,000 / $10,000 per BTC = 0.5 BTC
> 
>     2. **Liquidation Price Calculation**:
>        
>        * If maintenance_margin = 0%:
>          
>          * Liquidation Price = (price_opening * leverage) / (leverage - 1 + (maintenance_margin * leverage))
>            Liquidation Price = ($10,000 * 2) / (2 - 1 + (0.1 * 0))
>            Liquidation Price = $20,000 / (2 - 1)
>            Liquidation Price = $20,000 / 1
>            Liquidation Price = $20,000
>            Liquidation Price = $20,000 per BTC
>        * If `maintenance_margin = 10%`:
>          
>          * Liquidation Price = (price_opening * leverage) / (leverage - 1 + (maintenance_margin * leverage))
>            Liquidation Price = ($10,000 * 2) / (2 - 1 + (0.1 * 2))
>            Liquidation Price = $20,000 / (2 - 1 + 0.2)
>            Liquidation Price = $20,000 / (1 + 0.2)
>            Liquidation Price = $20,000 / 1.2
>            Liquidation Price ≈ $16,666.67
>          * --> meaning, the user gets liquidated earlier

Thanks for your detailed explanation and education on the different liquidation prices :). I've addressed your remarks in 3d28628f2de182c6fd8d6f7ff06b3b04f855d124

Please have another look.

File: coordinator/src/node/liquidated_positions.rs
Comment (holzeis) [reply coordinator/src/node/liquidated_positions.rs:177]: > Also we shouldn't need to recalculate the liquidation_price inside the function. We should have this in the position right?

yes, this has been addressed here 46eaa476cf0732d855f355b5a37243d04ecb63a9

Comment (holzeis) [reply coordinator/src/node/liquidated_positions.rs:181]: > 🔧 I find this function a bit too complex as it's checking both parties at the same time and returns a direction in which it's getting liquidated or None.
> 
> What's the reason for not having a function which only accepts one liquidation price?
> 
> It could look like this:
> 
> ```rust
> fn check_if_position_needs_to_get_liquidated(
>     trader_direction: Direction,
>     best_current_price: &Price,
>     liquidation_price: Decimal,
> ) -> bool {
> ...
> }
> ```
> 
> you can now call this twice, once for the coordinator and once for the trader.
> 
> The length of this function would then only be half what it's doing right now.

I refactored it quite significantly. Have a look here 0f1ca45f2371ae4711c0a4d97e0bd59e8ec92bf9

File: coordinator/src/node/liquidated_positions.rs
Comment (holzeis) [reply coordinator/src/node/liquidated_positions.rs]: done 8675cc129dbbfeb2419742181cbd20d655628859

File: crates/trade/src/cfd.rs
Comment (bonomat) [reply crates/trade/src/cfd.rs:53]: Done myself.

File: mobile/native/src/ln_dlc/mod.rs
Comment (bonomat) [inline mobile/native/src/ln_dlc/mod.rs:175]: Maybe good to log an error if this ever returns `None` or maybe even add an `expect` as it should always be set?
Otherwise this will just hide errors which are hard to find.

**Review by bonomat — APPROVED** [review]
Looks good. I've added a few comments which can be resolved later. I'll go ahead and rebase for you.

**Review by luckysori — COMMENTED** [review]
Half-way through, but it got merged.

File: coordinator/src/node/liquidated_positions.rs
Comment (luckysori) [inline coordinator/src/node/liquidated_positions.rs]: 🔧 It's a bit more readable to calculate both liquidation prices at the top.

File: coordinator/src/node/liquidated_positions.rs
Comment (luckysori) [inline coordinator/src/node/liquidated_positions.rs:185]: 🔧 We should avoid "Test-Driven Design Ossification"[^1].

In summary, we should not right tens of tests against the same API and instead call a data-driven test API that delegates to the API we want to test. I've used the `check` idiom [here](https://github.com/get10101/10101/pull/2352/files#diff-dc08eded476b7979ce9fa96345782e5d1c570f176643482d8baf65678b9e4477R947), for example.

[^1]: https://matklad.github.io/2021/05/31/how-to-test.html#Test-Driven-Design-Ossification.

File: coordinator/src/scheduler.rs
Comment (luckysori) [inline coordinator/src/scheduler.rs:256]: 📣 Looks to me like a risky `expect`.

File: coordinator/migrations/2024-04-03-105900_add_coordinator_liquidation_price/up.sql
Comment (luckysori) [reply coordinator/migrations/2024-04-03-105900_add_coordinator_liquidation_price/up.sql:5]: Hmmm, let's fix this though ❗

File: coordinator/src/node/liquidated_positions.rs
Comment (luckysori) [reply coordinator/src/node/liquidated_positions.rs:185]: This approach would have paid off twice before this PR even got merged lol.

File: coordinator/migrations/2024-04-02-110300_add_liquidation/down.sql
Comment (bonomat) [reply coordinator/migrations/2024-04-02-110300_add_liquidation/down.sql:2]: We need at least one query here or diesel will fail. I most of the time do a simple `select 1;`

File: coordinator/src/trade/mod.rs
Comment (luckysori) [inline coordinator/src/trade/mod.rs:533]: The name suggests that this is an `Amount`, but it's actually a percentage.

File: coordinator/src/node/liquidated_positions.rs
Comment (holzeis) [reply coordinator/src/node/liquidated_positions.rs:192]: yeah.. i thought about removing them, but they were already there so I guess they don't hurt 🤷‍♂️

File: crates/trade/src/cfd.rs
Comment (holzeis) [reply crates/trade/src/cfd.rs:53]: thanks

File: coordinator/src/notifications.rs
Comment (holzeis) [reply coordinator/src/notifications.rs:127]: We need the user to open the app to collaboratively execute the liquidation.

> Your position has been liquidated 📉😞 

Don't know if you'd have to be sad about it. It could also be that the counterparty got liquidated and you made a huge profit :)

> Open the app to try again!

I copied it from the expired position notification as its the same action the user needs to do, which is "open the app". I guess we should change it to this message :)

File: coordinator/src/node/liquidated_positions.rs
Comment (holzeis) [reply coordinator/src/node/liquidated_positions.rs:25]: Thanks for the reminder, this timeout only defines how long we try to collaborate with the user before we consider force closing the channel

File: coordinator/src/scheduler.rs
Comment (holzeis) [reply coordinator/src/scheduler.rs:256]: https://github.com/get10101/10101/pull/2371

File: coordinator/src/trade/mod.rs
Comment (holzeis) [reply coordinator/src/trade/mod.rs:533]: https://github.com/get10101/10101/pull/2372

File: coordinator/src/node/liquidated_positions.rs
Comment (holzeis) [reply coordinator/src/node/liquidated_positions.rs:202]: yes it is

Comment (holzeis) [reply coordinator/src/node/liquidated_positions.rs:202]: I removed the tests altogether since they aren't very valuable anymore.

https://github.com/get10101/10101/pull/2372

File: coordinator/src/notifications.rs
Comment (holzeis) [reply coordinator/src/notifications.rs:127]: https://github.com/get10101/10101/pull/2374

File: coordinator/src/node/liquidated_positions.rs
Comment (holzeis) [reply coordinator/src/node/liquidated_positions.rs]: changed in a later commit.

File: coordinator/src/node/liquidated_positions.rs
Comment (holzeis) [reply coordinator/src/node/liquidated_positions.rs:185]: The tests have been removed as after the refactoring they were basically only testing `>=` and `<=`.

File: coordinator/src/notifications.rs
Comment (luckysori) [reply coordinator/src/notifications.rs:127]: > Don't know if you'd have to be sad about it. It could also be that the counterparty got liquidated and you made a huge profit :)

I guess the message should depend on who got liquidated!

File: coordinator/src/node/liquidated_positions.rs
Comment (luckysori) [reply coordinator/src/node/liquidated_positions.rs:169]: This got simplified later, I think.

**Pull request merged**
